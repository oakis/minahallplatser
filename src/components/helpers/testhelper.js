export const navigation = {
    getParam: jest.fn(),
    setParams: jest.fn(),
    navigate: jest.fn(),
};
