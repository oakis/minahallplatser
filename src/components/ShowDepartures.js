import _ from 'lodash';
import { connect } from 'react-redux';
import React, { PureComponent } from 'react';
import { View, ScrollView, FlatList } from 'react-native';
import { Actions } from 'react-native-router-flux';
import fetch from 'react-native-cancelable-fetch';
import firebase from 'firebase';
import { renderHelpButton } from '../Router';
import { getDepartures, clearDepartures, clearErrors, favoriteLineToggle } from '../actions';
import { DepartureListItem, Spinner, Message, ListItemSeparator, Popup, Text } from './common';
import { updateStopsCount, track, incrementStopsOpened } from './helpers';
import { colors, component } from './style';

class ShowDepartures extends PureComponent {

	constructor(props) {
		super(props);
		this.state = {
			addFavorite: false,
			showHelp: false
		};
	}
	
	componentWillMount() {
		track('Page View', { Page: 'Departures', Stop: this.props.busStop });
		this.props.getDepartures({ id: this.props.id });
		updateStopsCount();
		const { currentUser } = firebase.auth();
		if (!currentUser.isAnonymous && _.includes(this.props.favoriteIds, this.props.id)) {
			incrementStopsOpened(currentUser.uid, this.props.id);
		}
	}

	componentDidMount() {
		this.startRefresh();
	}

	componentWillReceiveProps({ favorites, departures, timestamp }) {
		const favoritesUpdated = JSON.stringify(this.props.favorites) !== JSON.stringify(favorites);
		const departuresUpdated = JSON.stringify(this.props.departures) !== JSON.stringify(departures);
		if (favoritesUpdated) {
			this.populateFavorites(favorites);
		}
		if (departuresUpdated) {
			this.populateDepartures(departures);
		}
		if (this.props.timestamp !== timestamp) {
			Actions.refresh({ right: renderHelpButton(this) });
		}
	}

	componentWillUnmount() {
		clearInterval(this.interval);
		this.interval = null;
		this.props.clearDepartures();
		this.props.clearErrors();
		fetch.abort('getDepartures');
	}

	startRefresh() {
		const self = this;
		self.interval = setInterval(self.refresh.bind(self), 10000);
	}

	refresh() {
		Actions.refresh({ right: () => {
			return (
				<View style={{ flexDirection: 'row', justifyContent: 'center' }}>
					<Spinner color={colors.alternative} />
					{renderHelpButton(this)}
				</View>
			);
		} });
		this.props.getDepartures({ id: this.props.id });
	}

	populateFavorites(favorites) {
		window.log('Updated favorites:', favorites);
		this.props.favorites = favorites;
	}

	populateDepartures(departures) {
		window.log('Updated departures:', departures);
		this.props.departures = departures;
	}

	ListFooterComponent = () => {
		return (this.props.departures.length === 0 || this.props.favorites.length === 0) ? null : <View style={{ height: 5, backgroundColor: colors.primary }} />;
	}

	openPopup = () => {
		track('Show Help', { Page: 'Departures' });
		this.setState({
			showHelp: true
		});
	}

	renderDepartures = ({ item, index }) => {
		const { timeFormat } = this.props;
		const itemWithNewIndex = { ...item, index, timeFormat };
		return (
			<DepartureListItem
				item={itemWithNewIndex}
				onPress={() => {
					this.props.favoriteLineToggle(item);
					this.setState({ addFavorite: !this.state.addFavorite });
				}}
			/>
		);
	}

	renderPopup() {
		// const { imageWidth, imageHeight } = calcImageSize(0.85, metrics.margin.md * 2.5);
		return (
			<Popup
				onPress={() => this.setState({ showHelp: false })}
				isVisible={this.state.showHelp}
			>
				<Text style={component.popup.header}>När går nästa avgång?</Text>
				<Text style={component.popup.text}>Längst till höger på varje rad står det antal minuter kvar till nästa avgång samt avgången efter det. <Text style={{ fontStyle: 'italic' }}>Det går även att ändra avgångstiden till klockslag, och det gör du i menyn på startsidan.</Text></Text>
				{/* <Image style={component.popup.image} source={require('../assets/help/non-live.png')} style={{ width: imageWidth, height: imageHeight }} ImageResizeMode={'cover'} /> */}
				
				<Text style={component.popup.header}>Varför har tiden till nästa avgång ibland färg?</Text>
				<Text style={component.popup.text}>När en avgång snart ska gå från en hållplats så kommer alltid texten "<Text style={{ color: colors.danger }}>Nu</Text>" att visas med röd färg. Ibland kan man också se att en avgång har <Text style={{ color: colors.warning }}>orange</Text> text. Det kan t.ex betyda att en buss har tappat anslutningen med Västtrafik och inte längre är live. Tiden som visas då är ordinarie avgång enligt tidtabell.</Text>
				{/* <Image style={component.popup.image} source={require('../assets/help/non-live.png')} style={{ width: imageWidth, height: imageHeight }} ImageResizeMode={'cover'} /> */}
				
				<Text style={component.popup.header}>Hur sparar man en linje som favorit?</Text>
				<Text style={component.popup.text}>För att spara en linje så räcker det med att klicka på den, linjen kommer då hamna högst upp på alla hållplatser som den linjen kör.</Text>
				{/* <Image style={component.popup.image} source={require('../assets/help/non-live.png')} style={{ width: imageWidth, height: imageHeight }} ImageResizeMode={'cover'} /> */}
			</Popup>
		);
	}

	renderContent() {
		if (this.props.loading) {
			return (
				<Spinner
					size="large"
					color={colors.primary}
				/>
			);
		} else if (this.props.error) {
			return <Message type="warning" message={this.props.error} />;
		}

		return (
			<ScrollView>
				<FlatList
					data={this.props.favorites}
					renderItem={this.renderDepartures}
					keyExtractor={item => item.journeyid}
					ItemSeparatorComponent={ListItemSeparator}
					ListFooterComponent={this.ListFooterComponent}
					maxToRenderPerBatch={11}
					initialNumToRender={11}
					scrollEnabled={false}
					extraData={this.state}
				/>
				<FlatList
					data={this.props.departures}
					renderItem={this.renderDepartures}
					title={'departures'}
					keyExtractor={item => item.journeyid}
					ItemSeparatorComponent={ListItemSeparator}
					getItemLayout={(data, index) => (
						{ length: 51, offset: 51 * index, index }
					)}
					maxToRenderPerBatch={11}
					initialNumToRender={11}
					scrollEnabled={false}
					extraData={this.state}
				/>
			</ScrollView>
		);
	}

	render() {
		return (
			<View style={{ flex: 1 }}>
				{this.renderPopup()}
				{this.renderContent()}
			</View>
		);
	}
}

const MapStateToProps = (state) => {
	const { lines } = state.fav;
	const { loading, timestamp } = state.departures;
	let favorites = [];
	let departures = [];
	_.forEach(state.departures.departures, item => {
		const { sname, direction } = item;
		const departure = `${sname} ${direction}`;
		if (_.includes(lines, departure)) {
			favorites = [...favorites, item];
		} else {
			departures = [...departures, item];
		}
	});
	const { error } = state.errors;
	const { timeFormat } = state.settings;
	const favoriteDepartures = state.fav.favorites;
	const favoriteIds = _.map(favoriteDepartures, 'id');
	return { departures, loading, error, timestamp, favorites, timeFormat, favoriteIds };
};

export default connect(MapStateToProps,
	{ getDepartures, clearDepartures, clearErrors, favoriteLineToggle })(ShowDepartures);
