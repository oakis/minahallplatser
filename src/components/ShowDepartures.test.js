import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import ShowDepartures from './ShowDepartures';
import { navigation } from './helpers/testhelper.js';

const mockStore = configureMockStore();

const initialState = {
    fav: {
        lines: [
            '16 Högsbo',
        ],
    },
    departures: {
        departures: [
            {
                sname: '16X',
                direction: 'Högsbo',
                clockLeft: '16:00',
                clockNext: '16:05',
                timeLeft: '3',
                timeNext: '8',
                isLive: true,
                fgColor: '#eee',
                bgColor: '#000',
                type: 'BUS',
                via: 'via Centralstationen',
                track: 'A',
                night: false,
                accessibility: 'wheelChair',
                journeyid: '123',
            }
        ],
        loading: false,
    },
    errors: {
        error: '',
    },
    settings: {
        timeFormat: '',
    },
};

it('should match snapshot', () => {
    const wrapper = shallow(<ShowDepartures navigation={navigation} store={mockStore(initialState)} getDepartures={jest.fn()} incrementStopsOpened={jest.fn()} />).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('16 Högsbo should match 16X Högsbo', () => {
    const wrapper = shallow(<ShowDepartures navigation={navigation} store={mockStore(initialState)} />);
    const expected = '16X';
    const actual = wrapper.props().favorites[0].sname;
    expect(actual).toBe(expected);
});

it('16X Högsbo should match 16 Högsbo', () => {
    initialState.fav.lines = ['16X Högsbo'];
    initialState.departures.departures = [
        {
            sname: '16',
            direction: 'Högsbo'
        }
    ];
    const wrapper = shallow(<ShowDepartures navigation={navigation} store={mockStore(initialState)} />);
    const expected = '16';
    const actual = wrapper.props().favorites[0].sname;
    expect(actual).toBe(expected);
});
