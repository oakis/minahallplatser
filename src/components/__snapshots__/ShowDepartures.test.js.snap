// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <MiniMenu
    isVisible={false}
    items={
      Array [
        Object {
          "content": "Minuter",
          "onPress": [Function],
        },
        Object {
          "content": "Klockslag",
          "onPress": [Function],
        },
      ]
    }
    onClose={[Function]}
  />
  <MiniMenu
    isVisible={false}
    items={
      Array [
        Object {
          "content": "Ändra tidsformat",
          "icon": "access-time",
          "onPress": [Function],
        },
        Object {
          "content": "Hjälp",
          "icon": "help",
          "onPress": [Function],
        },
      ]
    }
    onClose={[Function]}
  />
  <Popup
    isVisible={false}
    onPress={[Function]}
  >
    <DefaultFont
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 5,
        }
      }
    >
      När går nästa avgång?
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      Längst till höger på varje rad står det antal minuter kvar till nästa avgång samt avgången efter det. 
      <DefaultFont
        style={
          Object {
            "fontStyle": "italic",
          }
        }
      >
        Det går även att ändra avgångstiden till klockslag, och det gör du i menyn på startsidan.
      </DefaultFont>
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 5,
        }
      }
    >
      Varför har tiden till nästa avgång ibland färg?
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      När en avgång snart ska gå från en hållplats så kommer alltid texten "
      <DefaultFont
        style={
          Object {
            "color": "#FF6961",
          }
        }
      >
        Nu
      </DefaultFont>
      " att visas med röd färg. Ibland kan man också se att en avgång har 
      <DefaultFont
        style={
          Object {
            "color": "#FFB347",
          }
        }
      >
        orange
      </DefaultFont>
       text. Det kan t.ex betyda att en buss har tappat anslutningen med Västtrafik och inte längre är live. Tiden som visas då är ordinarie avgång enligt tidtabell.
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 5,
        }
      }
    >
      Hur sparar man en linje som favorit?
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      För att spara en linje så räcker det med att klicka på den, linjen kommer då hamna högst upp på alla hållplatser som den linjen kör på. För att spara avgången som favorit för nuvarande hållplats så tryck på avgången och håll ner fingret i en halv sekund.
    </DefaultFont>
  </Popup>
  <ScrollView>
    <FlatList
      ItemSeparatorComponent={[Function]}
      ListFooterComponent={[Function]}
      data={
        Array [
          Object {
            "accessibility": "wheelChair",
            "bgColor": "#000",
            "clockLeft": "16:00",
            "clockNext": "16:05",
            "direction": "Högsbo",
            "fgColor": "#eee",
            "global": true,
            "isLive": true,
            "journeyid": "123",
            "local": false,
            "night": false,
            "sname": "16X",
            "timeLeft": "3",
            "timeNext": "8",
            "track": "A",
            "type": "BUS",
            "via": "via Centralstationen",
          },
        ]
      }
      extraData={
        Object {
          "init": true,
          "miniMenuOpen": false,
          "showHelp": false,
          "timeformatVisible": false,
        }
      }
      initialNumToRender={11}
      keyExtractor={[Function]}
      maxToRenderPerBatch={11}
      renderItem={[Function]}
      scrollEnabled={false}
    />
    <FlatList
      ItemSeparatorComponent={[Function]}
      data={Array []}
      extraData={
        Object {
          "init": true,
          "miniMenuOpen": false,
          "showHelp": false,
          "timeformatVisible": false,
        }
      }
      getItemLayout={[Function]}
      initialNumToRender={11}
      keyExtractor={[Function]}
      maxToRenderPerBatch={11}
      renderItem={[Function]}
      scrollEnabled={false}
    />
  </ScrollView>
</View>
`;
